packages/back-end/src/util/organization.util.ts

```diff
@@ -68,10 +71,77 @@ export function roleToPermissionMap(
   return permissionsObj;
 }
 
-export function getUserPermissions(
+function mergePermissions(
+  existingPermissions: UserPermission,
+  existingRole: MemberRole | undefined,
+  teamInfo: TeamInterface | ProjectMemberRole,
+  org: OrganizationInterface
+) {
+  const newPermissions = roleToPermissionMap(teamInfo.role, org);
+  for (const newPermission in newPermissions) {
+    // If the user doesn't have permission, but the team role does, add it
+    if (
+      !existingPermissions.permissions[newPermission as Permission] &&
+      newPermissions[newPermission as Permission]
+    ) {
+      existingPermissions.permissions[newPermission as Permission] =
+        newPermissions[newPermission as Permission];
+    }
+  }
+
+  if (!existingRole) {
+    existingPermissions.limitAccessByEnvironment =
+      teamInfo.limitAccessByEnvironment;
+    existingPermissions.environments = teamInfo.environments;
+  } else {
+    const roles = getRoles(org);
+    const newRoleIndex = roles.findIndex((r) => r.id === teamInfo.role);
+    const existingRoleIndex = roles.findIndex((r) => r.id === existingRole);
+
+    const existingRoleIsEngineerOrExperimenter =
+      existingRole === "engineer" || existingRole === "experimenter";
+
+    const newRoleIsEngineerOrExperimenter =
+      teamInfo.role === "engineer" || teamInfo.role === "experimenter";
+
+    if (
```

Comment:
Why do we only do the limit access by environment checks below if the role is `engineer` or `experimenter`? Is it because these two roles would be the only ones able to work with environments anyway? If so, I wonder if we can abstract that logic to somewhere else that is shared because I imagine if we add more roles that can work with environments we'd want to have a const or a function we can use to perform the check

Reply:
@msamper That is correct - only `engineer` and `experimenter` roles can be environment limited.

I did a bit of digging, and it looks like on the frontend we have a method `roleSupportsEnvLimit` - I've moved that to the shared folder, so we can now use that here (and continue using it on the frontend).

Good suggestion - I hated the pattern above. It reads much better now.